src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttribute.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Internal.Web.Utils;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class TagHelperAttribute : IReadOnlyTagHelperAttribute
+    {
+        public TagHelperAttribute()
+        {
+        }
+
+        public TagHelperAttribute(string name, object value)
+        {
+            Name = name;
+            Value = value;
+        }
+
+        public string Name { get; set; }
+
+        public object Value { get; set; }
+
+        public static implicit operator TagHelperAttribute(string value)
+        {
+            return new TagHelperAttribute
+            {
+                Value = value
+            };
+        }
+
+        public bool Equals(IReadOnlyTagHelperAttribute other)
+        {
+            return
+                other != null &&
+                string.Equals(Name, other.Name, StringComparison.OrdinalIgnoreCase) &&
+                Equals(Value, other.Value);
+        }
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as IReadOnlyTagHelperAttribute;
+
+            return Equals(other);
+        }
+
+        public override int GetHashCode()
+        {
+            return HashCodeCombiner
+                .Start()
+                .Add(Name, StringComparer.OrdinalIgnoreCase)
+                .Add(Value)
+                .CombinedHash;
```

Comment:
reread the `object.GetHashCode()` docs for other reasons and would strongly recommend removing both `Name` and `Value` from this calculation.  what I've been doing in a separate cleanup where no immutable properties are involved in an overridden `Equals()` comparisons is:
``` c#
  private static readonly int TypeHashCode = typeof(TagHelperAttribute).GetHashCode();
  ...
  public override int GetHashCode()
  {
    return TypeHashCode;
  }
```

suggest the same for `CaseSensitiveTagHelperAttributeComparer` even though that's test code.

TL;DR;
- using `Name` or `Value` here is incorrect because they're mutable. if they are changed after a `TagHelperAttribute` is added to a hash tree, the instance will be lost.
- using `base.GetHashCode()` is incorrect because when `Equals()` is `true`, `GetHashCode()` should return the same value.  (fine for `GetHashCode()` to return the same value for instances for which `Equals()` is `false`.)